jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/RabitTracker.scala

```diff
@@ -0,0 +1,158 @@
+ */
+
+/**
+  * Pure Scala implementation of the DMLC tracker using Akka, ported from the
+  * original Python implementation. This implementation gets rid of the Python
+  * dependency in the Java version (RabitTracker.java) and therefore is less
+  * suspectible to Python-related issues.
+  */
+
+package ml.dmlc.xgboost4j.scala
+
+import java.net.{InetAddress, InetSocketAddress}
+import java.util.concurrent.TimeoutException
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future, Promise}
+import scala.util.{Failure, Random, Success, Try}
+
+import akka.actor.ActorSystem
+import akka.pattern.ask
+
+import ml.dmlc.xgboost4j.java.IRabitTracker
+import ml.dmlc.xgboost4j.scala.handler.RabitTrackerHandler
+
+/**
+  * Synchronous tracker class that mimics the Java API.
+  *
+  * {{{
+  *   import scala.concurrent.duration._
+  *
+  *   val tracker = new RabitTracker(32)
+  *   // allow up to 1 minute for the tracker to bind to a socket address.
+  *   tracker.start(1 minute)
+  *   /* ...
+  *      launching workers in parallel
+  *      ...
+  *   */
+  *
+  *   // wait for worker execution up to 6 hours.
+  *   // providing a finite timeout prevents a long-running task from hanging forever in
+  *   // catastrophic events, like the loss of an executor during model training.
+  *   tracker.waitFor(6 hours)
+  * }}}
+  *
+  * @param numWorkers number of distributed workers, each corresponding to a Spark task.
+  * @param port The minimum port that the tracker binds to. If port is omitted, or given as None,
+  *             a random ephemeral port is chosen at runtime.
+  * @param maxPortTrials The maximum number of trials of socket binding, by sequentially
+  *        increasing port.
+  * @param workerConnectionTimeout The timeout for awaiting connections from worker nodes.
+  *        When used in Spark applications, because all Spark transformations are lazily executed,
+  *        the I/O time for loading RDDs from external sources (local disk, HDFS, S3 etc.) must be
+  *        taken in account. If the timeout value is too small, the Rabit workers will likely time
+  *        out before connecting, due to the overhead of loading RDDs.
+  *        Using a finite timeout is encouraged, as it prevents the tracker (thus the Spark driver
+  *        running it) from hanging indefinitely due to worker connection issues (e.g. firewall).
+  */
+class RabitTracker(numWorkers: Int, port: Option[Int] = None,
+                   maxPortTrials: Int = 1000,
+                   workerConnectionTimeout: Duration = 10 minutes) extends IRabitTracker {
+  import scala.collection.JavaConverters._
+
+  require(numWorkers >=1, "numWorkers must be greater than or equal to one (1).")
+
+  val system = ActorSystem.create("RabitTracker")
+  val handler = system.actorOf(RabitTrackerHandler.props(
+    numWorkers, maxPortTrials, workerConnectionTimeout), "Handler")
+  implicit val askTimeout: akka.util.Timeout = akka.util.Timeout(1 second)
```

Comment:
Yes, in an implicit fashion for the ask function (`?`).
Although it is a good idea to unify all of the these wait/ask timeouts.